auth support for tls endpoint

crazy-max opened the pull request:
~needs #163~

Adds authentication support for `tls://` endpoint (and docs to also set up SSH auth).

To ease the integration in a workflow, environment variables are used to avoid registered inputs in the action for each node (see #165). You need to add the following environment variables to set up TLS authentication with the BuildKit client certificates:

* `BUILDER_NODE_<idx>_AUTH_TLS_CACERT`
* `BUILDER_NODE_<idx>_AUTH_TLS_CERT`
* `BUILDER_NODE_<idx>_AUTH_TLS_KEY`

Where `<idx>` is the position of the node in the list of nodes. The index should always be `0` at the moment as we don't support (yet) appending new nodes with this action.

jedevc commented on docs/advanced/auth.md:
```diff
@@ -0,0 +1,101 @@
+# Authentication to a remote node
```
Not sure about the page title, since the remote driver works without authentication - we probably don't want to promote using unauthenticated TCP connections, but it's allowed, and for a setup that runs buildkitd and buildctl on the same CI machine, it seems like a reasonable deployment?

Unrelated, but we should also mention the `docker-container://` scheme here, happy to do that as a follow-up later.

crazy-max replied:
> but it's allowed, and for a setup that runs buildkitd and buildctl on the same CI machine, it seems like a reasonable deployment?

Yes but don't think users want to set up a local buildkit instance on an ephemeral GitHub Runner? I think they prefer to use this action to create a BuildKit container.

> Unrelated, but we should also mention the `docker-container://` scheme here, happy to do that as a follow-up later.

Yes sounds good. I guess it would use a remote context to connect to a remote docker engine instance?

jedevc reviewed (commented):
Nice :tada:

Not so sure about the `env` scheme, especially since we support other arguments using `with` - is there a reason we couldn't put node authentication details there, using a `nodes` field? It feels a little inconsistent to have to track multiple arguments over different fields.

Potentially we could use that for specifying `platforms` per node as well? I think I'm imagining something that's like the refactoring of output in #162, but for the inputs instead.

crazy-max commented:
> Not so sure about the `env` scheme, especially since we support other arguments using `with`- is there a reason we couldn't put node authentication details there, using a `nodes` field? It feels a little inconsistent to have to track multiple arguments over different fields.

Don't think there is a better alternative. Using `with`, we can't define arbitrary inputs. They must be defined in `action.yml` and with #165, user could have an arbitrary list of nodes and therefore as many authentication entries. See https://github.com/crazy-max/docker-setup-buildx-action/blob/append/docs/advanced/append-nodes.md

> Potentially we could use that for specifying `platforms` per node as well? I think I'm imagining something that's like the refactoring of output in #162, but for the inputs instead.

We could change the format but ssh keys and client certs are multi line values and we can't read secrets directly in the action with the current API :disappointed:

crazy-max commented:
@tonistiigi As discussed, I removed the SSH authentication handling in the action as it would tamper the local machine when modifying the SSH configuration. Also updated the docs to show how to set up SSH auth using a third-party action instead.
